### PR TITLE
feat(trin execution): added support for dao fork + evm tracer

### PR DESCRIPTION
**Pull request by KolbyML**
### What was wrong?

Our state execution didn't support the DAO fork. Which caused us to generate an incorrect state root and the client to crash

Our code wasn't setup to easily provide traces to debug with

### How was it fixed?

By draining all the specified accounts and moving the balances to the specified account as specified in EIP-779

also by adding a cli config for tracing blocks

### CODE REVIEW

File: trin-execution/src/execution.rs [hunk 1]
```diff
@@ -213,6 +221,61 @@ impl State {
             }
         }
 
+        // check if dao fork, if it is drain accounts and transfer it to beneficiary
+        if block_tuple.header.header.number == get_spec_block_number(SpecId::DAO_FORK) {
+            let mut drained_balance_sum = U256::ZERO;
```
Comment (morph-dev) [inline trin-execution/src/execution.rs:226]: This should be in `dao_fork.rs`. I think you will just have to pass `&mut self.database`, which is fine.

File: trin-execution/src/execution.rs [hunk 2]
```diff
@@ -255,11 +319,32 @@ impl State {
...
                     Transaction::Blob(tx) => tx.modify(block_number, tx_env),
                 }
-            })
-            .with_external_context(NoOpInspector)
-            .append_handler_register(inspector_handle_register)
-            .build()
-            .transact()?)
+            });
+
+        Ok(
```
Comment (morph-dev) [inline trin-execution/src/execution.rs:324]: I think something like this is much nicer:
```rust
let mut evm_builder = Evm::builder()
    ...; // part from above

if self.config.block_to_trace.should_trace(block_number) {
  ...
  evm_builder = evm_builder.with_external_context(
      TracerEip3155::new(Box::new(output_file)));
} else {
  evm_builder = evm_builder.with_external_context(NoOpInspector);
}

Ok(evm_builder
    .append_handler_register(inspector_handle_register)
    .build()
    .transact()?)
```

File: trin-execution/src/types/block_to_trace.rs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+impl FromStr for BlockToTrace {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "none" => Ok(BlockToTrace::None),
+            "all" => Ok(BlockToTrace::All),
+            val => {
+                let index = val
```
Comment (morph-dev) [inline trin-execution/src/types/block_to_trace.rs:29]: Use [`str.split_once()`](https://doc.rust-lang.org/std/primitive.str.html#method.split_once). For example:

```rust
match s.split_once(':') {
  Some(("block", value)) => {
      let block_number = value
          .parse()
          .map_err(|err| format!("Invalid block number: {err}"))?;
      Ok(BlockToTrace::Block(block_number))
  },
  _ => Err(format!("Invalid block trace argument: {s}"))
}
```

**Review by morph-dev — APPROVED** [review]
Only few small comments
:rocket:

File: trin-execution/src/execution.rs (hunk 2 above)
Comment (KolbyML) [reply trin-execution/src/execution.rs:324]: I tried earlier, but I get this 
image

Comment (KolbyML) [reply trin-execution/src/execution.rs:324]: I tried your example quite a few things and I am not sure this example or idea is possible. I originally tried your example when I first tried implementing this a few days ago, but I couldn't get it to compile

File: trin-execution/src/config.rs [hunk 4]
```diff
@@ -1,15 +1,28 @@
...
 impl Default for StateConfig {
     fn default() -> Self {
         Self {
             cache_contract_storage_changes: false,
+            block_to_trace: BlockToTrace::None,
+        }
+    }
+}
+
+impl From<TrinExecutionConfig> for StateConfig {
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply trin-execution/src/config.rs:21]: We can change it in the future if it makes sense, but for now we don't need a reference so I will leave it for now

**Pull request merged**